fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryRecursivePlan.java

```diff
@@ -0,0 +1,299 @@
+        if (this == otherExpression) {
+            return true;
+        }
+        if (getClass() != otherExpression.getClass()) {
+            return false;
+        }
+        return semanticEqualsForResults(otherExpression, aliasMap);
+    }
+
+    @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
+    @Override
+    public boolean equals(final Object other) {
+        return structuralEquals(other);
+    }
+
+    @Override
+    public int hashCode() {
+        return structuralHashCode();
+    }
+
+    @Override
+    public int hashCodeWithoutChildren() {
+        return Objects.hash(getResultValue()); // priorValueCorrelation _is_ a child correlation due to the recursive nature of this plan.
+    }
+
+    @Nonnull
+    @Override
+    public RelationalExpression translateCorrelations(@Nonnull final TranslationMap translationMap, final boolean shouldSimplifyValues,
+                                                      @Nonnull final List<? extends Quantifier> translatedQuantifiers) {
+        Verify.verify(translatedQuantifiers.size() == 2);
+        Verify.verify(!translationMap.containsSourceAlias(priorValueCorrelation));
+        return new RecordQueryRecursivePlan(
+                translatedQuantifiers.get(0).narrow(Quantifier.Physical.class),
+                translatedQuantifiers.get(1).narrow(Quantifier.Physical.class),
+                priorValueCorrelation
+        );
+    }
+
+    @Override
+    public void logPlanStructure(StoreTimer timer) {
+    }
+
+    @Override
+    public int getComplexity() {
+        return rootQuantifier.getRangesOverPlan().getComplexity() * childQuantifier.getRangesOverPlan().getComplexity();
+    }
+
+    @Override
+    public int planHash(@Nonnull final PlanHashMode mode) {
+        switch (mode.getKind()) {
+            case LEGACY:
+            case FOR_CONTINUATION:
+                return PlanHashable.objectsPlanHash(mode, BASE_HASH, getChildren(), getResultValue());
+            default:
+                throw new UnsupportedOperationException("Hash kind " + mode.name() + " is not supported");
+        }
+    }
+
+    @Nonnull
+    @Override
+    public List<? extends Quantifier> getQuantifiers() {
+        return ImmutableList.of(rootQuantifier, childQuantifier);
+    }
+
+    @Nonnull
+    @Override
+    public PlannerGraph rewritePlannerGraph(@Nonnull final List<? extends PlannerGraph> childGraphs) {
+        return PlannerGraph.fromNodeAndChildGraphs(
+                new PlannerGraph.OperatorNodeWithInfo(this,
+                        NodeInfo.NESTED_LOOP_JOIN_OPERATOR,
+                        ImmutableList.of("RECURSIVE {{expr}}"),
+                        ImmutableMap.of("expr", Attribute.gml(getResultValue().toString()))),
+                childGraphs,
+                getQuantifiers());
+    }
+
+    @Nonnull
+    @Override
+    public PRecordQueryRecursivePlan toProto(@Nonnull final PlanSerializationContext serializationContext) {
```

Comment:
Please extend the naming changes to the protobuf classes as well.

Reply:
Done.